sos/collector/clusters/pacemaker.py

```diff
@@ -18,42 +22,80 @@ class pacemaker(Cluster):
+        ('offline', True, 'Collect nodes listed as offline'),
+        ('only-corosync', False, 'Only use corosync.conf to enumerate nodes')
     ]
 
     def get_nodes(self):
-        self.res = self.exec_primary_cmd('pcs status')
-        if self.res['status'] != 0:
-            self.log_error('Cluster status could not be determined. Is the '
-                           'cluster running on this node?')
-            return []
-        if 'node names do not match' in self.res['output']:
-            self.log_warn('Warning: node name mismatch reported. Attempts to '
-                          'connect to some nodes may fail.\n')
-        return self.parse_pcs_output()
-
-    def parse_pcs_output(self):
-        nodes = []
-        if self.get_option('online'):
-            nodes += self.get_online_nodes()
-        if self.get_option('offline'):
-            nodes += self.get_offline_nodes()
-        return nodes
-
-    def get_online_nodes(self):
-        for line in self.res['output'].splitlines():
-            if line.startswith('Online:'):
-                nodes = line.split('[')[1].split(']')[0]
-                return [n for n in nodes.split(' ') if n]
-
-    def get_offline_nodes(self):
-        offline = []
-        for line in self.res['output'].splitlines():
-            if line.startswith('Node') and line.endswith('(offline)'):
-                offline.append(line.split()[1].replace(':', ''))
-            if line.startswith('OFFLINE:'):
-                nodes = line.split('[')[1].split(']')[0]
-                offline.extend([n for n in nodes.split(' ') if n])
-        return offline
+        self.nodes = []
+        # try crm_mon first
+        try:
+            if not self.get_option('only-corosync'):
+                try:
+                    self.get_nodes_from_crm()
+                except Exception as err:
+                    self.log_warn("Could not parse crm_mon output: %s" % err)
+            if not self.nodes:
+                # fallback to corosync.conf, in case the node we're inspecting
+                # is offline from the cluster
+                self.get_nodes_from_corosync()
+        except Exception as err:
+            self.log_error("Could not determine nodes from cluster: %s" % err)
+
+        _shorts = [n for n in self.nodes if '.' not in n]
+        if _shorts:
+            self.log_warn(
+                "WARNING: Node addresses '%s' may not resolve locally if you "
+                "are not running on a node in the cluster. Try using option "
+                "'-c pacemaker.only-corosync' if these connections fail."
+                % ','.join(_shorts)
+            )
+        return self.nodes
+
+    def get_nodes_from_crm(self):
+        """
+        Try to parse crm_mon output for node list and status.
+        """
+        xmlopt = '--output-as=xml'
+        # older pacemaker had a different option for xml output
+        _ver = self.exec_primary_cmd('crm_mon --version')
```

Comment:
You might want to use XML for this too, just in case.
```
# crm_mon --version --output-as=xml
<pacemaker-result api-version="2.12" request="crm_mon --version --output-as=xml">
  <version program="Pacemaker" version="2.1.0-8.el8" author="Andrew Beekhof" build="7c3f660707" features="agent-manpages cibsecrets compat-2.0 corosync-ge-2 default-concurrent-fencing default-sbd-sync generated-manpages monotonic nagios ncurses remote systemd"/>
  <status code="0" message="OK"/>
</pacemaker-result>
```

Reply:
Never mind, I just realize that would be circular.

So, maybe a `try`/`except` block instead of a version check.

Reply:
Tl;dr, probably fine.

I was concerned about futureproofing, not pythonicity. The text output format of `crm_mon --version` is subject to change (though it doesn't seem particularly likely to change). If it changes, then we may fail to parse out the version that we need later to determine `--output-as=xml` versus `--as-xml`.

A solution would be to get XML output from `crm_mon --version`, but that becomes circular... you'd need the version in order to know which "as xml" option to use in `crm_mon --version`.

That's why I suggested skipping the whole version check and just using try/except.

In practice, it's unlikely to matter. If the `crm_mon --version` format has changed, then that means we're on a newer version and so we won't hit the `version < 2.3` condition to choose `--as-xml`. The only way it would matter is if `crm_mon --version` format changes and then the `--output-as=xml` syntax also changes.